Sources/LiveViewNative/Property Wrappers/LiveBinding.swift

```diff
@@ -0,0 +1,215 @@
+///
+/// ### Server Support
+/// As live bindings are two-way mechanism, they are not implemented purely on the client, but rather require server-side support.
+///
+/// The server sends an event on mount with all of the binding values, sends an event when a binding is changed, and handles an even when a binding is changed on the client. These are all provided by the [`live_view_native_swift_ui`](https://github.com/liveviewnative/live_view_native_swift_ui) Elixir package.
+///
+/// The `bindings` macro takes a keyword list of bindings and their initial values and automatically generates all the necessary supporting code.
+/// The Elixir value used for the initial binding value must be serializable as JSON.
+///
+/// ```elixir
+/// defmodule AppWeb.TestLive do
+///     use AppWeb, :live_view
+///     use LiveViewNativeSwiftUi.Bindings
+///
+///     bindings(toggle_binding: false, alert_shown: false)
+/// end
+/// ```
+///
+/// Then, in the template, the same name is given as the value of a binding attribute:
+///
+/// ```html
+/// <my-toggle is-on="toggle_binding" />
+///
+/// <button modifiers='[{"type": "my_alert", "is_active": "alert_shown"}]'>
+///     <text>Present alert</text>
+/// </button>
+/// ```
+///
+/// ### Client Usage
+/// To use this property wrapper, the wrapped type must impelement the `Codable` protocol to define how values are serialized over the network.
+///
+/// ```swift
+/// struct MyToggle: View {
+///     @LiveBinding(attribute: "is-on") private var isOn: Bool
+///
+///     var body: some View {
+///         Toggle(isOn: $isOn) {
+///             Text("My Toggle")
+///         }
+///     }
+/// }
+///
+/// struct MyAlertModifier: ViewModifier, Decodable {
+///     @LiveBinding private var isActive: Bool
+///
+///     init(from decoder: Decoder) throws {
+///         let container = try decoder.container(keyedBy: CodingKeys.self)
+///         self._isActive = try container.decode(LiveBinding.self, forKey: .isActive)
+///     }
+///
+///     func body(content: Content) -> some View {
+///         content
+///             .alert("Hello", isPresented: $isActive) {
+///                 Button("OK") {}
+///             }
+///     }
+///
+///     enum CodingKeys: String, CodingKey {
+///         case isActive = "is_active"
+///     }
+/// }
+/// ```
+@propertyWrapper
+public struct LiveBinding<Value: Codable>: Decodable {
+    @StateObject private var data = Data()
+    
+    @ObservedElement private var element
+    @Environment(\.coordinatorEnvironment) private var coordinator
+    @EnvironmentObject private var liveViewModel: LiveViewModel
+    
+    private let bindingNameSource: BindingNameSource
+    private var bindingName: String {
+        switch bindingNameSource {
+        case .fixed(let name):
+            return name
+        case .attribute(let attr):
+            guard let name = element.attributeValue(for: attr) else {
+                fatalError("@LiveBinding missing binding name for \(attr)")
```

Comment:
`LiveBinding` still cannot be optional. This could be changed fairly easily so that it can accept an optional default value which is then used if the attribute is not found. Then in the `nonmutating set` call `data.objectWillChange.send()` to ensure the value is not updating locally.

<details>
<summary>Here's the changes I made to `LiveBinding` to support this, so that it could work well with List's `selection` attribute.</summary>

```swift
@propertyWrapper
public struct LiveBinding<Value: Codable>: Decodable {
    @StateObject private var data = Data()
    
    @ObservedElement private var element
    @Environment(\.coordinatorEnvironment) private var coordinator
    @EnvironmentObject private var liveViewModel: LiveViewModel
    
    private let defaultValue: Value?
    
    private let bindingNameSource: BindingNameSource
    private var bindingName: String? {
        switch bindingNameSource {
        case .fixed(let name):
            return name
        case .attribute(let attr):
            return element.attributeValue(for: attr)
        }
    }
    
    var isBound: Bool {
        switch bindingNameSource {
        case .fixed(_):
            return true
        case .attribute(let attr):
            return element.attribute(named: attr) != nil
        }
    }
    
    /// Creates a `LiveBinding` property wrapper that uses the binding in the given attribute.
    ///
    /// This initializer should be used when the live binding is used as part of an element.
    /// See ``LiveBinding`` for a discussion of how the underlying binding name is determined and an example of this usage.
    public init(wrappedValue: Value? = nil, attribute: AttributeName) {
        self.bindingNameSource = .attribute(attribute)
        self.defaultValue = wrappedValue
    }
    
    /// Creates a `LiveBinding` by decoding its name from a container.
    ///
    /// This initializer should be used when the live binding is used as part of a view modifier.
    /// See ``LiveBinding`` for an example of how this is used.
    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let name = try container.decode(String.self)
        self.bindingNameSource = .fixed(name)
        self.defaultValue = nil
    }
    
    /// The value of the binding.
    ///
    /// Setting this property will send an event to the server to update its value as well.
    public var wrappedValue: Value {
        get {
            guard let bindingName else { return defaultValue! }
            return data.getValue(from: liveViewModel, bindingName: bindingName)
        }
        nonmutating set {
            guard let bindingName else {
                data.objectWillChange.send()
                return
            }
            // update the local value
            data.mode = .local(newValue)
            // update the view model, which will send an update to the backend
            let encoder = FragmentEncoder()
            // todo: if encoding fails, what should happen?
            try! newValue.encode(to: encoder)
            liveViewModel.setBinding(bindingName, to: encoder.unwrap() as Any)
        }
    }
    
    /// A SwiftUI `Binding` that provides read and write access to the underlying data.
    ///
    /// Access this binding with the dollar sign prefix. If a property is declared as `@LiveBinding(name: "value") var value`, then the projected value binding can be access as `$value`.
    public var projectedValue: Binding<Value> {
        Binding {
            return self.wrappedValue
        } set: { newValue in
            self.wrappedValue = newValue
        }
    }
}
```

</details>

Reply:
Ah, I see, you want to be able to use the LiveBinding even when a binding name isn't specified by the server. I was think of it as, in that case, you would have a separate State property. But I think you're right, it makes sense to fold that into LiveBinding.

Reply:
Or actually, is that not the case? Your code implies that LiveBinding is read-only and always returns the default value when a binding name isn't set.

I think a read/write local-only mode makes sense (I ran into needing that with TextEditor, and did this which isn't nice:
```swift
@State private var localIsFindPresented = false
@LiveBinding(attribute: "find-presented") private var liveIsFindPresented: Bool
var findPresented: Binding<Bool> { _liveIsFindPresented.isBound ? $liveIsFindPresented : $localIsFindPresented }
```

In that case, read/write access is necessary because passing `Binding.constant(false)` to the `findNavigator` modifier would, I believe, prevent it from being shown, which is not what we want if the binding name is omitted.